components/support/rc_crypto/src/contentsignature.rs

```diff
@@ -0,0 +1,305 @@
+            Ok(v) => v,
+            Err(_) => return Err(ErrorKind::RootHashFormatError(input.to_string()).into()),
+        };
+        result.extend(byte);
+    }
+
+    Ok(result)
+}
+
+fn split_pem(pem_content: &[u8]) -> Result<Vec<Vec<u8>>> {
+    let pem_str = match str::from_utf8(pem_content) {
+        Ok(v) => v,
+        Err(e) => {
+            return Err(ErrorKind::PEMFormatError(e.to_string()).into());
+        }
+    };
+
+    let pem_lines = pem_str.split('\n');
+
+    let mut blocks: Vec<Vec<u8>> = vec![];
+    let mut block: Vec<u8> = vec![];
+    let mut read = false;
+    for line in pem_lines {
+        if line.contains("-----BEGIN CERTIFICATE") {
+            read = true;
+        } else if line.contains("-----END CERTIFICATE") {
+            read = false;
+            let decoded = match base64::decode(&block) {
+                Ok(v) => v,
+                Err(e) => return Err(ErrorKind::PEMFormatError(e.to_string()).into()),
+            };
+            blocks.push(decoded);
+            block.clear();
+        } else if read {
+            block.extend_from_slice(&line.as_bytes());
+        }
+    }
+    if read {
+        return Err(ErrorKind::PEMFormatError("Missing end header".into()).into());
+    }
+    if blocks.is_empty() {
+        return Err(ErrorKind::PEMFormatError("Missing PEM data".into()).into());
+    }
+
+    Ok(blocks)
+}
+
+pub fn verify(
+    input: &[u8],
+    signature: &[u8],
+    pem_bytes: &[u8],
+    seconds_since_epoch: u64,
+    root_sha256_hash: &str,
+    hostname: &str,
+) -> Result<()> {
+    let certificates = split_pem(&pem_bytes)?;
+
+    let mut certificates_slices: Vec<&[u8]> = vec![];
+    for certificate in &certificates {
+        certificates_slices.push(certificate);
+    }
+
+    let root_hash_bytes = decode_root_hash(&root_sha256_hash)?;
+
+    nss::pkixc::verify_code_signing_certificate_chain(
+        certificates_slices,
+        seconds_since_epoch,
+        &root_hash_bytes,
+        hostname,
+    )
+    .map_err(|err| match err.kind() {
+        nss::ErrorKind::CertificateIssuerError => ErrorKind::CertificateIssuerError,
+        nss::ErrorKind::CertificateExpiredError => ErrorKind::CertificateExpiredError,
+        nss::ErrorKind::CertificateSubjectError => ErrorKind::CertificateSubjectError,
+        _ => ErrorKind::CertificateChainError(err.to_string()),
+    })?;
+
+    let leaf_cert = certificates.first().unwrap(); // PEM parse fails if len == 0.
+
+    let public_key_bytes = match nss::cert::extract_public_key(&leaf_cert) {
+        Ok(bytes) => bytes,
+        Err(err) => return Err(ErrorKind::CertificateContentError(err.to_string()).into()),
+    };
+
+    let signature_bytes = match base64::decode_config(&signature, base64::URL_SAFE) {
+        Ok(b) => b,
+        Err(err) => return Err(ErrorKind::SignatureContentError(err.to_string()).into()),
+    };
+
+    // Since signature is NIST384p / secp384r1, we can perform a few safety checks.
```

Comment:
Out of curiosity, what would happen if we *didn't* do these checks?

Reply:
I added a comment below, but basically the signature verification would fail for a number of different reasons. Doing these checks here help us being more explicit about the reason of failure.